src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+                SetConditionalState(UnreachableState(), this.State);
+            }
+            else if ((object)node.Type == null || node.Type.SpecialType != SpecialType.System_Boolean)
+            {
+                // a dynamic type or a type with operator true/false
+                Unsplit();
             }
 
-            VisitStatement(catchBlock.Body);
+            Split();
         }
 
-        protected virtual void VisitFinallyBlock(BoundStatement finallyBlock, ref TLocalState unsetInFinally)
+        /// <summary>
+        /// Visit a general expression, where we will only need to determine if variables are
+        /// assigned (or not). That is, we will not be needing AssignedWhenTrue and
+        /// AssignedWhenFalse.
+        /// </summary>
+        /// <param name="node"></param>
+        protected BoundNode VisitRvalue(BoundExpression node)
         {
-            VisitStatement(finallyBlock); // this should generate no pending branches
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            var result = Visit(node);
+            Unsplit();
+            return result;
         }
 
-        #endregion TryStatements
+        /// <summary>
+        /// Visit a statement.
+        /// </summary>
+        [DebuggerHidden]
+        protected virtual void VisitStatement(BoundStatement statement)
+        {
+            Debug.Assert(!_trackExceptions || this.PendingBranches.Count > 0 && this.PendingBranches[0].Branch == null);
+            Visit(statement);
+            Debug.Assert(!this.IsConditionalState);
+        }
+
+        protected static bool IsConstantTrue(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.True;
+        }
+
+        protected static bool IsConstantFalse(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.False;
+        }
+
+        protected static bool IsConstantNull(BoundExpression node)
+        {
+            return node.ConstantValue == ConstantValue.Null;
+        }
+
+        /// <summary>
+        /// Called at the point in a loop where the backwards branch would go to.
+        /// </summary>
+        private void LoopHead(BoundLoopStatement node)
+        {
+            TLocalState previousState;
+            if (_loopHeadState.TryGetValue(node, out previousState))
+            {
+                Join(ref this.State, ref previousState);
+            }
+
+            _loopHeadState[node] = this.State.Clone();
+        }
+
+        /// <summary>
+        /// Called at the point in a loop where the backward branch is placed.
+        /// </summary>
+        private void LoopTail(BoundLoopStatement node)
+        {
+            var oldState = _loopHeadState[node];
+            if (Join(ref oldState, ref this.State))
+            {
+                _loopHeadState[node] = oldState;
+                this.stateChangedAfterUse = true;
```

Comment:
1. why true?
2. why is oldState not being cloned here?
3. why is oldState being pulled out, but then jammed back in?

Reply:
> why true? Because Join returned true, indicating a nontrivial change to oldState.

Note that this ties into the previous comment that `bool Join` needed to document the return value.  That 'true' indicates 'non-trivial change' wasn't clear :)

2. it would be helpful to have more 'proof' around these decisions.  For example, how do we know that it is only modified here?  Similarly, for the code that is like this, but does clone the state, why it needs to do that.  i.e. what it's expectations are for who else might be modifying things.  From my reading of this code, these are the questions that leap out at me as being unclear as to how i would make a correctness decision.  Having documentation helps clear that up.

3. "Because this is being called at the point in a loop where a backwards branch appears. Just like it says in the method comment."  The comment does nto feel sufficient in terms of explaining the behavior.  For example, imagine the code looked like this:

```c#
            if (Join(ref oldState, ref this.State))
            {
                this.stateChangedAfterUse = true;
```

And you asked "why are we not adding the state back in?".  If i responded with "Because this is being called at the point in a loop where a backwards branch appears. Just like it says in the method comment." that wouldn't really explain why that code would be correct, or answer why it was correct to behave this way.

--

Aside, the code might be clearer as:  `if (Join(ref _loopHeadState[node], ref this.State)) ...`

That might make it clearer that this was en emplaced update of that specific state. 

Just my 2c on clarity reading through this.  I'm fairly familiar with the flow analysis stuff having had to walk through it many times.  BUt even so, some of the code can be rather cryptic.